### PR TITLE
feat(24.04): add python3 slice

### PR DESCRIPTION
**Pull request by linostar**
# Proposed changes

Add python3 slice, which depends on python3.12 slices and add a couple of symlinks over that.

## Related issues/PRs

N/A

### Forward porting

N/A

## Checklist

* [x] I have read the [contributing guidelines](
https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md)
* [x] I have tested my changes ([see how](https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md#7-test-your-slices-before-opening-a-pr))
* [x] I have already submitted the [CLA form](
https://ubuntu.com/legal/contributors/agreement)

## Additional Context

N/A

### CODE REVIEW

File: slices/python3.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+  standard:
+    essential:
+      - python3.12_standard
+
+  utils:
+    essential:
+      - python3.12_utils
+    contents:
+      /usr/bin/pdb3:
```
Comment (clay-lake) [inline slices/python3.yaml:19]: Do these files need to be symlinked to their python 3.12 alternatives in python3.12_utils?
ex. /usr/bin/pdb3 -> /usr/bin/pdb3.12

File: slices/python3.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package: python3
+
+essential:
+  - python3_copyright
+
+slices:
```
Comment (clay-lake) [inline slices/python3.yaml:6]: Should we modify or add a slice to link python3.12 to python3 and possibly python?

File: slices/python3.yaml (hunk 1 above)
Comment (linostar) [reply slices/python3.yaml:19]: They are already symlinks when installed from python3 package, and they are linked properly (pdb3 linked to pdb3.12, etc).

**Review by zhijie-yang — APPROVED** [review]
Looks good to me, thanks!

File: slices/python3.yaml (hunk 2 above)
Comment (linostar) [reply slices/python3.yaml:6]: A link from python to python3.12 is not created by the postinst script of python3 package, so I didn't include that in this slice. Besides, linking python3 to a specific python3.x version will cause conflicts when there are multiple python3.x packages in the same release (e.g. python3.10 and python3.11 in 22.04).

**Review by clay-lake — APPROVED** [review]
LGTM Thanks!

File: slices/python3.yaml (hunk 2 above)
Comment (cjdcordeiro) [reply slices/python3.yaml:6]: I also think we should have the `python3` symlink, as that might be the most useful thing about this SDF.

It is true that it isn't set by this pkg. The symlink is created by `python3-minimal` which is a dependency that is missing from this PR.

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]

File: slices/python3.yaml (hunk 2 above)
Comment (cjdcordeiro) [reply slices/python3.yaml:6]: The latest change artificially creates the python3 symlink, but I don't think that's the right way to do it. The symlink is created by python3-minimal which is a dependency that should be included in this PR.

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]

**Comment by linostar** [comment]
@cjdcordeiro python3-minimal is already included as an indirect dependency (see my comment above).
I'll move the symlink to python3-minimal SDF instead.

File: slices/python3.yaml (hunk 2 above)
Comment (linostar) [reply slices/python3.yaml:6]: Done.

File: slices/libpython3-stdlib.yaml [hunk 3]
```diff
@@ -0,0 +1,110 @@
+package: libpython3-stdlib
+
```
Comment (cjdcordeiro) [inline slices/libpython3-stdlib.yaml:2]: missing the copyright

File: slices/python3-minimal.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+slices:
+  bins:
+    essential:
+      - python3-minimal_config
+      - python3-minimal_modules
+      - python3.12-minimal_bins
+    contents:
+      /usr/bin/py3clean:
+      /usr/bin/py3compile:
+      /usr/bin/py3versions:
```
Comment (cjdcordeiro) [inline slices/python3-minimal.yaml:15]: do we need these? 

If we want to keep the `bins` slice, all of these must be tested.

In the end, we only care about the python3 symlink right? If so, I'd create a dedicated slice for it (and test it).

File: slices/python3-minimal.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      /usr/bin/python3:
+
+  config:
+    contents:
+      /usr/share/python3/debian_defaults:
+
+  modules:
```
Comment (cjdcordeiro) [inline slices/python3-minimal.yaml:22]: why are these called "modules"?

Does `py3versions.py` need `debpython`? The former is related to the bin py3versions symlink, so they could stay together.

File: slices/python3.yaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      - libpython3-stdlib_language
+      - libpython3-stdlib_markup-tools
+      - libpython3-stdlib_multimedia
+      - libpython3-stdlib_net-data
+      - libpython3-stdlib_numeric
+      - libpython3-stdlib_osx-support
+      - libpython3-stdlib_pydoc
+      - libpython3-stdlib_text
+      - libpython3-stdlib_unix
+      - python3-minimal_bins
```
Comment (cjdcordeiro) [inline slices/python3.yaml:42]: It is more maintainable if you replace `libpython3-stdlib_core` and `libpython3-stdlib_core` by `python3_core`

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]

File: slices/python3-minimal.yaml (hunk 5 above)
Comment (linostar) [reply slices/python3-minimal.yaml]: debian_defaults needed by py3versions. The others are needed by py3compile/py3clean. Added more slices for those.

File: slices/python3-minimal.yaml (hunk 4 above)
Comment (linostar) [reply slices/python3-minimal.yaml]: Added more slices.

File: slices/libpython3-stdlib.yaml (hunk 3 above)
Comment (linostar) [reply slices/libpython3-stdlib.yaml:2]: Added.

File: slices/python3-minimal.yaml (hunk 4 above)
Comment (cjdcordeiro) [reply slices/python3-minimal.yaml]: Thanks.

We need to be consistent about `bins`.

As per other slices, `bins` typically deliver all binaries.

In this case, we want the `/usr/bin/python3` symlink, so we should have a dedicated slice for it.

You've created dedicated slices for the other bins too, which is ok.

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
Need to address https://github.com/canonical/chisel-releases/pull/387#discussion_r1853655947 and the failing tests

**Review by cjdcordeiro — APPROVED** [review]
thanks for the changes. lgtm

**Pull request merged**